pkg/ingester/ingester.go

```diff
@@ -217,6 +217,55 @@ func (i *Ingester) starting(ctx context.Context) error {
 	return nil
 }
 
+// NewForFlusher constructs a new Ingester to be used by flusher target.
+func NewForFlusher(cfg Config, clientConfig client.Config, chunkStore ChunkStore, registerer prometheus.Registerer) (*Ingester, error) {
+	if cfg.ingesterClientFactory == nil {
+		cfg.ingesterClientFactory = client.MakeIngesterClient
+	}
+
+	i := &Ingester{
+		cfg:          cfg,
+		clientConfig: clientConfig,
+		metrics:      newIngesterMetrics(registerer, true),
+		chunkStore:   chunkStore,
+		flushQueues:  make([]*util.PriorityQueue, cfg.ConcurrentFlushes),
+	}
+
+	level.Info(util.Logger).Log("msg", "recovering from WAL")
+
+	// We recover from WAL always.
+	start := time.Now()
+	if err := recoverFromWAL(i); err != nil {
+		level.Error(util.Logger).Log("msg", "failed to recover from WAL", "time", time.Since(start).String())
+		return nil, err
+	}
+	elapsed := time.Since(start)
+
+	level.Info(util.Logger).Log("msg", "recovery from WAL completed", "time", elapsed.String())
+	i.metrics.walReplayDuration.Set(elapsed.Seconds())
+
+	// Should be a noop WAL.
+	cfg.WALConfig.WALEnabled = false
+	var err error
+	i.wal, err = newWAL(cfg.WALConfig, i.userStates.cp)
+	if err != nil {
+		return nil, err
+	}
+
+	i.Service = services.NewBasicService(i.startingForFlusher, i.loop, i.stopping)
+	return i, nil
+}
+
+func (i *Ingester) startingForFlusher(ctx context.Context) error {
+	i.flushQueuesDone.Add(i.cfg.ConcurrentFlushes)
+	for j := 0; j < i.cfg.ConcurrentFlushes; j++ {
+		i.flushQueues[j] = util.NewPriorityQueue(i.metrics.flushQueueLength)
+		go i.flushLoop(j)
+	}
+
```

Comment:
This same code is in `starting`, would it make sense to extract into new method (eg. startFlushLoops)?